build_wheels.py

```diff
@@ -0,0 +1,227 @@
+from packaging.requirements import Requirement
+
+import constants
+
+idf_constraints_url = 'https://dl.espressif.com/dl/esp-idf/espidf.constraints.'
+
+
+def check_response(response: requests.Response, warning: str, exit_on_wrong: bool = False) -> bool:
+    """Print warning or exit the script when response code is not correct
+
+    Args:
+        response (requests.Response): response to check status code
+        warning (str): warning to be printed
+        exit (bool, optional): if True rises SystemExit when status code != 200 ... Defaults to False.
+
+    Raises:
+        SystemExit: if status code != 200 and exit is True
+
+    Returns:
+        bool: True if status code 200
+    """
+    if response.status_code != 200:
+        if exit_on_wrong:
+            raise SystemExit(f'{warning}\n{response.text}')
+        else:
+            print(warning, '\n')
+    return True
+
+##############################################################
+# --- Lists of branches and constraints files to be downloaded requirements for --- #
+# IDF branches are automatically fetched from github, specified in constants.py
+# For constraints file the version is resolved from IDF release branch
+# The constraints file for master branch is considered to be the latest release version
+# incremented by 1 in minor field (e.g when last release is v5.2 the constrains file for master is v5.3)
+# if there is no such a version, the last release branch constraints file will be used
+# This can be overwritten by setting the variable 'AUTO_VERSION_CONSTRAINS_FILE' to False in constants.py
+# and 'MASTER_CONSTRAINS_FILE' is then taken for master branch
+
+
+# ESP-IDF branches list
+def fetch_idf_branches() -> list:
+    """Fetch IDF branches from URL specified in constants.py
+
+    Returns:
+        list: list of all IDF branches fetched from URL
+    """
+    res = requests.get(constants.IDF_BRANCHES_URL)
+    if check_response(res, 'Failed to fetch ESP-IDF branches.', True):
+        return [branch['name'] for branch in res.json()]
+    return []
+
+
+def get_used_idf_branches(idf_repo_branches: list) -> list:
+    """Take only IDF master and release branches, only equal or grater version specified in constants.py
+
+    Args:
+        idf_repo_branches (list): IDF fetched branches to select from
+
+    Returns:
+        list: IDF branches to download requirements for
+    """
+    idf_branches = []
+    for branch in idf_repo_branches:
+        try:
+            idf_major, idf_minor = re.match(r'release/v(\d+)\.(\d+)', branch).groups()  # type: ignore
+            if (int(idf_major), int(idf_minor)) >= (constants.MIN_IDF_MAJOR_VERSION, constants.MIN_IDF_MINOR_VERSION):
+                idf_branches.append(branch)
+        except AttributeError:
+            continue
+    idf_branches.append('master')
+    return idf_branches
+
+
+# Constraints files versions list
+def get_constraints_versions(idf_branches: list) -> list:
```

Comment:
This nesting functions look suspicious 🤨  ....

Perhaps you could consider implement it with guard/short-circuit style like this:
```py
def get_constraints_versions(idf_branches: List[str]) -> List[str]:
    """From desired branches passed in get constraints files versions list."""
    idf_constraints = []

    for branch in idf_branches:
        # Handle release branches
        if 'release/' in branch:
            idf_constraints.append(branch.split('release/')[1])
            continue

        # Handle master branch
        if not constants.AUTO_VERSION_CONSTRAINS_FILE:
            idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
            continue

        # If there are no previous constraints, use the master constraints file
        if not idf_constraints:
            idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
            continue

        # Parse the last version number and increment the minor version
        match = re.match(r'v(\d+)\.(\d+)', idf_constraints[-1])
        if not match:
            # No valid version found in the last constraint, use the master constraints file
            idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
            continue  # Skip to the next iteration

        # A valid version was found, increment the minor version
        last_major, last_minor = map(int, match.groups())
        next_minor_version = f'v{last_major}.{last_minor + 1}'
        response = requests.get(f'{IDF_CONSTRAINTS_URL}{next_minor_version}.txt', timeout=10)
        # Append the next minor version if it exists, otherwise append the last known good constraints
        idf_constraints.append(next_minor_version if response.status_code == 200 else idf_constraints[-1])

    return idf_constraints
```

This is more flat and easy to read and maintain.